[doc] Add project internals guide

stephenfin opened the pull request:
This is essentially a refactoring of the huge CONTRIBUTING.rst document. This document is split into multiple separate documents based on the Django documentation. Some minor modifications are made to the documents in the process.

stephenfin committed:
docs: Tweaks to code of conduct

- Replace unicode quotes with ascii, since Sphinx will automatically
  translate these for us
- Move the header to the root document
- Use an admonition instead of a header for the attribution section

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
docs: Expand organization document

Provide some guidance on how one can become a core developer and explain
the different types of contribution one can make.

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented on doc/internals/contributing.rst:
```diff
@@ -0,0 +1,309 @@
+======================
```
Could you separate the commits into moving (renaming) and modification? AFAIK, It keeps the change history of the file. Additionally, it helps to read changes from individual commit logs for me.

I know it does not help file separation. But it would be better to keep history even only for "contributing to sphinx" part.

stephenfin replied:
I've tried to do this. Hopefully it's okay now?

tk0miya replied:
No, it's not good. It seems you did moving CONTRIBUTING.rst to doc/internals/contributing.rst and adding new CONTRIBUTING.rst in the same commit. It seems git considers it large deletion of CONTRIBUTING.rst and creation of doc/internals/contributing.rst.
https://github.com/sphinx-doc/sphinx/pull/7750/commits/7e8ae1127577ca60d6ce98b9c958b6ddb554a44e

As a result, the history of doc/internals/contributing.rst has been cleared.
https://github.com/stephenfin/sphinx/blame/doc/internals-section/doc/internals/contributing.rst


Please separate renaming and other actions.
For example, this branch commits 3 times; 1) moving 2) adding 3) modifying. It allows to keep history.
https://github.com/sphinx-doc/sphinx/compare/3.x...tk0miya:moving_test
https://github.com/tk0miya/sphinx/blame/moving_test/doc/internals/contributing.rst

stephenfin replied:
Okay, I think I have this done now.

tk0miya commented on doc/internals/index.rst:
```diff
@@ -0,0 +1,17 @@
+   release-process
+   organization
+   authors
+   changes
```
I don't think changes is not a part of "internal". Why do you move this here?

stephenfin replied:
Fair point. Moved back to root.

stephenfin committed:
docs: Add "project internals" guide

This is modelled off Django's guide of the same name [1]. In summary,
this will contain most of the information previously found in the
CONTRIBUTING document. For now, we move just the code of conduct and
authors documents.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
docs: Add organization overview

Provide an introduction to how the Sphinx team is organized. This will
be expanded upon in a future change.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
docs: Add release process document

This is based on content extracted from the CONTRIBUTING document.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
docs: Rework contributing document, re-add CONTRIBUTING

Reorder this to flow a little better and re-add the CONTRIBUTING
document, which now simply points to the main documentation.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin commented:
Rebased onto `3.x`

stephenfin committed:
docs: Add contributing document

This completes the migration from the CONTRIBUTING document to the new
in-tree contributing guide. The CONTRIBUTING document itself is
temporarily removed so that git will correctly mark the document as
having been moved. A future change will re-add this document as well as
rework the new contributing document.

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented:
Thank you for updates. I'll merge this after a moment.

tk0miya commented:
Update redirect settings on RTD. merging now.
